crates/ra_macros/src/mbe_expander.rs

```diff
@@ -0,0 +1,169 @@
+use rustc_hash::FxHashMap;
+use smol_str::SmolStr;
+
+use crate::{mbe, tt, tt_cursor::TtCursor};
+
+pub fn exapnd(rules: &mbe::MacroRules, input: &tt::Subtree) -> Option<tt::Subtree> {
+    rules.rules.iter().find_map(|it| expand_rule(it, input))
+}
+
+fn expand_rule(rule: &mbe::Rule, input: &tt::Subtree) -> Option<tt::Subtree> {
+    let mut input = TtCursor::new(input);
+    let bindings = match_lhs(&rule.lhs, &mut input)?;
+    expand_subtree(&rule.rhs, &bindings, &mut Vec::new())
+}
+
+#[derive(Debug, Default)]
+struct Bindings {
+    inner: FxHashMap<SmolStr, Binding>,
+}
+
+#[derive(Debug)]
+enum Binding {
+    Simple(tt::TokenTree),
+    Nested(Vec<Binding>),
+}
+
+impl Bindings {
+    fn get(&self, name: &SmolStr, nesting: &[usize]) -> Option<&tt::TokenTree> {
+        let mut b = self.inner.get(name)?;
+        for &idx in nesting.iter() {
+            b = match b {
+                Binding::Simple(_) => break,
+                Binding::Nested(bs) => bs.get(idx)?,
+            };
+        }
+        match b {
+            Binding::Simple(it) => Some(it),
+            Binding::Nested(_) => None,
+        }
+    }
+    fn push_nested(&mut self, nested: Bindings) -> Option<()> {
+        for (key, value) in nested.inner {
+            if !self.inner.contains_key(&key) {
```

Comment:
Is there an `entry` api we can use instead?

Reply:
probably, though we'd still need that match agains `Nested`...